### PR TITLE
Simplify tests suite, remove spy_phase and related functions

### PR DESCRIPTION
**Pull request by lebrice**
# Description

This PR addresses #884 , by removing the pretty "spy_phase" stuff from the test suite. This makes tests a lot more transparent.

# Changes

- All tests of a BaseAlgorithmTests sublass are now parametrized to run through each of the different test phases.
  - A new `first_phase_only` decorator can be used to disable this behaviour.
  - Running tests in all phases has actually been useful: I've identified a few subtle bugs related to using set_state and get_state after the initial number of trials have been observed.
- a new `phase` fixture can be requested to get the current `TestPhase`.
- `BaseAlgoTests.create_algo` now creates the algo, and makes it observe the right number of trials for the current phase before returning it.

This PR also fixes the bugs that were highlighted by this change in the test suite, for all the algorithms.

## Tests
- [X] I added corresponding tests for bug fixes and new features. If possible, the tests fail without the changes
- [ ] All new and existing tests are passing (`$ tox -e py38`; replace `38` by your Python version if necessary)

## Documentation
- [ ] I have updated the relevant documentation related to my changes

## Quality
- [X] I have read the [CONTRIBUTING](https://github.com/Epistimio/orion/blob/develop/CONTRIBUTING.md) doc
- [X] My commits messages follow [this format](https://chris.beams.io/posts/git-commit/)
- [ ] My code follows the style guidelines (`$ tox -e lint`)

### CODE REVIEW

File: src/orion/algo/hyperband.py
Comment (bouthilx) [inline src/orion/algo/hyperband.py:304]: Is this one of the subtil bugs you referred to in the PR description?

File: src/orion/algo/pbt/pbt.py
Comment (bouthilx) [inline src/orion/algo/pbt/pbt.py:373]: Is this bug fixed by this if-else?

File: tests/unittests/algo/test_evolution_es.py
Comment (bouthilx) [inline tests/unittests/algo/test_evolution_es.py:546]: That's concerning. Did you found out why? Please let me know if you would need help to investigate this.

File: tests/unittests/algo/test_evolution_es.py
Comment (bouthilx) [inline tests/unittests/algo/test_evolution_es.py:508]: It's not really a cleanup, it's more like a generic test that is added at the end of the tests. But it is not really necessary to run at end of all tests and could actually be a separate test on its own, using the phase parametrization too.

File: src/orion/testing/algo.py
Comment (bouthilx) [inline src/orion/testing/algo.py:157]: That will pollute the test list with a lot of skips. :thinking: It would nice if the phases could be used when a method has the argument `phase` or something like that, like the usual way of using pytest fixtures.

File: src/orion/testing/algo.py
Comment (bouthilx) [inline src/orion/testing/algo.py:221]:
```suggestion
                    f"Test isn't configured properly: max_trials ({cls.max_trials}) is smaller than "
```

File: src/orion/testing/algo.py
Comment (bouthilx) [inline src/orion/testing/algo.py:231]: So far the algorithms are not cached to be reused across the tests for a given phase right?

File: src/orion/testing/algo.py
Comment (bouthilx) [inline src/orion/testing/algo.py:419]: I think this test will not work properly now since the new `create_algo` implementation is executing `algo.seed_rng`. Because of this we cannot test that the algo was seeded properly at instantiation.

File: src/orion/testing/algo.py
Comment (bouthilx) [inline src/orion/testing/algo.py:448]: But the algo should be reproducible anyhow, so the first suggested trial should be the same. Along the same lines, we assume the initial state to be the same, and to the be same for both after sampling the first trial.

File: tests/unittests/algo/test_evolution_es.py
Comment (bouthilx) [reply tests/unittests/algo/test_evolution_es.py:546]: (Also this should probably be another issue/PR, not work for this one)

**Comment by bouthilx** [comment]
Thanks a lot for this PR! This is a great improvement on the parametrization of the phases for the algorithm test-suite. My only main concern is with the automatic parametrization of all tests which can lead to lots of skips (configuration test, is_done test, cardinality test, convergence test). If we could why a simple way to mark methods so that they are parametrized for first phase only that would be great.

File: src/orion/algo/hyperband.py
Comment (lebrice) [reply src/orion/algo/hyperband.py:304]: Yes!

Comment (bouthilx) [reply src/orion/algo/hyperband.py:304]: Ok great, good catch! :)

File: src/orion/algo/pbt/pbt.py
Comment (lebrice) [reply src/orion/algo/pbt/pbt.py:373]: Yes, the if/else fixes it.
Can I leave this note ?

File: tests/unittests/algo/test_evolution_es.py
Comment (lebrice) [reply tests/unittests/algo/test_evolution_es.py:546]: Created a new issue for this: https://github.com/Epistimio/orion/issues/890

I did not find out why, but I know that the trial's parents has a different ID as well (IDK if that's useful).

File: tests/unittests/algo/test_evolution_es.py
Comment (lebrice) [reply tests/unittests/algo/test_evolution_es.py:508]: I agree: What I was suggesting was to move this testing logic after the `yield` statement inside a fixture, in other words, in the "cleanup" phase, not that this is cleanup code.

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:157]: Hmm. Not sure I agree.
I think all tests of a BaseMethodTests subclass should be run in all phases by default.
I see marking a test as `first_phase_only` a bit like adding an `xfail` on all the other phases. It's not something you should do often.

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:231]: Yeah, they aren't. There would be a problem with the ordering of the parameterization from the fixture. I think I left a note somewhere about this.
I'll remove this TODO, it's outdated anyhow.

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:419]: `create_algo` only calls `seed_rng` if a seed is passed as a parameter.

Comment (lebrice) [reply src/orion/testing/algo.py:419]: Does that address your concern?

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:448]: This doesn't just test the initial state anymore, it tests at each phase. This condition is correct, right?
I'll rephrase the note a bit.

File: src/orion/algo/pbt/pbt.py
Comment (lebrice) [reply src/orion/algo/pbt/pbt.py:373]: (since I don't think this is particularly clean, it's more like a patch for a bug)

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py]: Rephrased the note in ecba557eb89cddf71eeb540e0d4af43f94e1841a

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:228]: Removed the todo in e456637ec38ea396be7f1f3713c52dbe7e65d63d

**Comment by lebrice** [comment]
> If we could why a simple way to mark methods so that they are parametrized for first phase only that would be great.
 
I wouldn't know how to do this currently, since the `phase` fixture is already parameterized by the `params` when it actually gets access to the `request` fixture (where it could inspect `fixturenames` for instance).

I think it's fine to expect all tests to be parametrized to run on each phase by default (even configuration tests, for instance).

Also, The docs job is failing for a weird reason, some kind of internal pytests thing. I'll try to sort that out.

File: src/orion/algo/pbt/pbt.py
Comment (bouthilx) [reply src/orion/algo/pbt/pbt.py:373]: Leaving the note is fine, but it should be more clear that this is a hacky bug fix, rather than a bug to be fixed.

File: tests/unittests/algo/test_evolution_es.py
Comment (bouthilx) [reply tests/unittests/algo/test_evolution_es.py:546]: Thanks!

File: src/orion/testing/algo.py
Comment (bouthilx) [reply src/orion/testing/algo.py:158]: All tests that are for argument values, configuration, convergence and is_done are not meant to be parametrized with phases. Depending on the algorithm that can be from about 1/5 to 1/3 of the tests.

File: src/orion/testing/algo.py
Comment (bouthilx) [reply src/orion/testing/algo.py:430]: See in this case it will clearly call seed_rng, wheras this test is for seeding durint init, and thus seed_rng should not be called. It's problematic.

File: src/orion/testing/algo.py
Comment (bouthilx) [reply src/orion/testing/algo.py]: Oh, I did not realize the changes to the test due to the modification of create_algo. So previously we would have `algo` updated up to the phase, but the new_algo would be at initial phase (num=0). I think it would be best to keep it this way, to make sure algorithms can have their state updated from different phases instead of always just one sample away. For this `create_algo` would need to take as argument whether it should be from scratch or at current phase.

Comment (bouthilx) [reply src/orion/testing/algo.py]: Just to make sure we are not reinserting a blind spot here, do you remember which test made you find the bug in hyperband when you were modifying the test-suite?

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:430]: I don't think that the "call `seed_rng` internally if seed is passed to constructor" behaviour is something we want to enforce, since algos don't have to take a seed as a constructor parameter.

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py]: > I think it would be best to keep it this way

You mean the way I have it here? I'm confused.

Comment (lebrice) [reply src/orion/testing/algo.py]: > For this create_algo would need to take as argument whether it should be from scratch or at current phase.

Why do you need this, if the test is parametrized to run on all phases? Would you mind explaining again?

Comment (lebrice) [reply src/orion/testing/algo.py]: > Just to make sure we are not reinserting a blind spot here, do you remember which test made you find the bug in hyperband when you were modifying the test-suite?

Which bug? Sorry I'm a bit slow this morning I guess

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py:430]: I guess I could skip the test if the algo under test doesn't have a `seed` constructor parameter..

File: src/orion/testing/algo.py
Comment (bouthilx) [inline src/orion/testing/algo.py:162]:
```suggestion
        """Returns the number of trials in the given phase. """
```

File: src/orion/testing/algo.py
Comment (bouthilx) [reply src/orion/testing/algo.py:158]: For convergence and is_done it should only be last phase IMO. Arguments and configuration should be first phase only.

Skipping tests is fine in general but skipping systematically 1/3 of the tests is a lot. It's not a big issue, but it will pollute the pytest report. If you can find an easy fix that would be great but otherwise we'll just leave it like that.

Comment (lebrice) [reply src/orion/testing/algo.py:192]: Added a first_phase_only and last_phase_only fixture (and make the mentioned tests use it) in 1ce04a9a8b14b2193000c9222880632cacf4e7b2

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py]: Fixed in 1ce04a9a8b14b2193000c9222880632cacf4e7b2

File: src/orion/testing/algo.py
Comment (lebrice) [reply src/orion/testing/algo.py]: Fixed in c88d490e382d116532dd4735c7723d3479ec7aea

I still don't quite recall what bug you're referring to. :(

**Comment by bouthilx** [comment]
Is this PR waiting for a final review from my side? Sorry I lost track in the past 2 weeks.

**Comment by lebrice** [comment]
> Is this PR waiting for a final review from my side? Sorry I lost track in the past 2 weeks.

Yep. I'll fix the merge conflict tomorrow, but apart from that, it's ready to go IMO

File: src/orion/algo/pbt/pbt.py
Comment (bouthilx) [reply src/orion/algo/pbt/pbt.py]: Still unaddressed. It's noting critical, but it's confusing.

File: src/orion/testing/algo.py
Comment (bouthilx) [reply src/orion/testing/algo.py]: So far yes they need to: https://github.com/Epistimio/orion/blob/develop/src/orion/algo/gridsearch.py#L116. GridSearch is the only situation where we don't use the seed so far.

File: src/orion/testing/algo.py
Comment (bouthilx) [reply src/orion/testing/algo.py]: It's ok, I don't think you told me exactly what it was back then. Let's hope it is fixed properly. I'm it is. :)

**Review by bouthilx — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
